packages/verkle/src/verkleTree.ts

```diff
@@ -213,11 +215,168 @@ export class VerkleTree {
+        this.verkleCrypto
+      )
+      this.DEBUG && this.debug(`Creating new leaf node at stem: ${bytesToHex(stem)}`, ['PUT'])
+    }
+    // Update value in leaf node and push to putStack
+    leafNode.setValue(suffix, value)
+    this.DEBUG &&
+      this.debug(
+        `Updating value for suffix: ${suffix} at leaf node with stem: ${bytesToHex(stem)}`,
+        ['PUT']
+      )
+    putStack.push([leafNode.hash(), leafNode])
+
+    // `path` is the path to the last node pushed to the `putStack`
+    let lastPath = leafNode.stem
+
+    // Step 2) Determine if a new internal node is needed
+    if (foundPath.stack.length > 1) {
+      // Only insert new internal node if we have more than 1 node in the path
+      // since a single node indicates only the root node is in the path
+      const nearestNodeTuple = foundPath.stack.pop()!
+      const nearestNode = nearestNodeTuple[0]
+      lastPath = nearestNodeTuple[1]
+      const updatedParentTuple = this.updateParent(leafNode, nearestNode, lastPath)
+      putStack.push([updatedParentTuple.node.hash(), updatedParentTuple.node])
+      lastPath = updatedParentTuple.lastPath
+
+      // Step 3) Walk up trie and update child references in parent internal nodes
+      while (foundPath.stack.length > 1) {
+        const [nextNode, nextPath] = foundPath.stack.pop()! as [InternalNode, Uint8Array]
+        // Compute the child index to be updated on `nextNode`
+        const childIndex = lastPath[matchingBytesLength(lastPath, nextPath)]
+        // Update child reference
+        nextNode.setChild(childIndex, {
+          commitment: putStack[putStack.length - 1][1].commitment,
+          path: lastPath,
+        })
+        this.DEBUG &&
+          this.debug(
+            `Updating child reference for node with path: ${bytesToHex(
+              lastPath
+            )} at index ${childIndex} in internal node at path ${bytesToHex(nextPath)}`,
+            ['PUT']
+          )
+        // Hold onto `path` to current node for updating next parent node child index
+        lastPath = nextPath
+        putStack.push([nextNode.hash(), nextNode])
+      }
+    }
+
+    // Step 4) Update root node
+    const rootNode = foundPath.stack.pop()![0] as InternalNode
+    rootNode.setChild(stem[0], {
+      commitment: putStack[putStack.length - 1][1].commitment,
+      path: lastPath,
+    })
+    this.root(this.verkleCrypto.serializeCommitment(rootNode.commitment))
+    this.DEBUG &&
+      this.debug(
+        `Updating child reference for node with path: ${bytesToHex(lastPath)} at index ${
+          lastPath[0]
+        } in root node`,
+        ['PUT']
+      )
+    this.DEBUG && this.debug(`Updating root node hash to ${bytesToHex(this._root)}`, ['PUT'])
+    putStack.push([this._root, rootNode])
+    await this.saveStack(putStack)
   }
 
+  /**
+   * Helper method for updating or creating the parent internal node for a given leaf node
+   * @param leafNode the reference leaf node
```

Comment:
By reference leaf node I assume you mean the "new leaf node". Do you think that "newLeafNode/new leaf node" would be a clearer naming?

Reply:
Hmm, it could be a new leaf node but it could also just be an existing leaf node where we've updated a value.  In this instance, we wouldn't be creating a new internal node above but would still update the parent.  Maybe I can call it "child leaf node"?

Reply:
Ah that's fair. I'm not actually sure I find child leaf node clearer. I think the comment makes it clear enough, thanks for addressing.